nan_persistent_12_inl.h

```diff
@@ -81,47 +85,80 @@ class Global : public v8::Global<T> {
     v8::PersistentBase<T>::Reset(v8::Isolate::GetCurrent(), other);
   }
 
+  Global Pass() { return static_cast<Global&&>(*this); }  // NOLINT(build/c++11)
+
   template<typename P>
   inline void SetWeak(
     P *parameter
     , typename WeakCallbackInfo<P>::Callback callback
     , WeakCallbackType type) {
-    reinterpret_cast<Persistent<T>*>(this)->SetWeak(
-        parameter, callback, type);
+    static_cast<Persistent<T>*>(static_cast<v8::PersistentBase<T>*>(this))->
+        SetWeak(parameter, callback, type);
   }
 };
 #else
 template<typename T>
 class Global : public v8::UniquePersistent<T> {
+  struct RValue {
+    inline explicit RValue(v8::UniquePersistent<T> *obj) : object_(obj) {}
+    v8::UniquePersistent<T> *object_;
+  };
+
  public:
   inline Global() : v8::UniquePersistent<T>() {}
 
-  template<typename S> inline Global(v8::Local<S> that) :
+  template<typename S> inline explicit Global(v8::Local<S> that) :
     v8::UniquePersistent<T>(v8::Isolate::GetCurrent(), that) {}
 
   template<typename S>
-  inline Global(const v8::PersistentBase<S> &that) :
+  inline explicit Global(const v8::PersistentBase<S> &that) :
       v8::UniquePersistent<S>(v8::Isolate::GetCurrent(), that) {}
 
   inline void Reset() { v8::PersistentBase<T>::Reset(); }
 
-  template <typename S>
+  template<typename S>
   inline void Reset(const v8::Local<S> &other) {
     v8::PersistentBase<T>::Reset(v8::Isolate::GetCurrent(), other);
   }
 
-  template <typename S>
+  template<typename S>
   inline void Reset(const v8::PersistentBase<S> &other) {
     v8::PersistentBase<T>::Reset(v8::Isolate::GetCurrent(), other);
   }
 
+#if NODE_MODULE_VERSION == NODE_0_12_MODULE_VERSION
+  inline void Empty() {
+    static_cast<v8::Persistent<T>*>(static_cast<v8::PersistentBase<T>*>(this))->
+        ClearAndLeak();
+  }
+#endif
+
+  inline Global(RValue rvalue) {  // NOLINT(runtime/explicit)
+    std::memcpy(this, rvalue.object_, sizeof (*rvalue.object_));
+# if NODE_MODULE_VERSION > NODE_0_12_MODULE_VERSION
+    rvalue.object_->Empty();
+# else
+    static_cast<v8::Persistent<T>*>(static_cast<v8::PersistentBase<T>*>(
+        rvalue.object_))->ClearAndLeak();
+# endif
+  }
+
+  template<typename S>
+  inline Global &operator=(v8::UniquePersistent<S> other) {
+    return static_cast<Global&>(v8::UniquePersistent<S>::operator=(other));
+  }
+
+  inline operator RValue() { return RValue(this); }
```

Comment:
It is necessary for move emulation. The compiler sees the (private) copy constructor of non-const ref. It does not work with a temporary, so it looks for other options. It finds the `Global(RValue)` constructor and looks for a conversion to `RValue`. The conversion is provided by the conversion operator, so this succeeds. The `Global(RValue)` constructor then moves ownership of the encapsulated value to `this`.